doc/ne.1
.TH NE 1 "by Sebastiano Vigna and Todd M. Lewis" "ne" \" -*- nroff -*-
.SH NAME
ne \- A nice editor
.SH SYNOPSIS
.B ne [options] files
.TP
Options:
[\-\-help]
[\-\-]
[+[N[,M]]]
[\-\-binary]
[\-\-read-only]
[\-\-utf8]
[\-\-no\-utf8]
[\-\-ansi]
[\-\-no\-ansi]
[\-\-no\-config]
[\-\-no\-syntax]
[\-\-prefs ext]
[\-\-keys key\-configuration\-file]
[\-\-menus menu\-configuration\-file]
[\-\-macro macro\-file]

.SH DESCRIPTION
\fBne\fR is a free text editor that runs on (hopefully almost) any UN*X
machine.  \fBne\fR is easy to use for the beginner, but powerful and fully
configurable for the wizard, and most sparing in its resource usage.

This documentation is incomplete.  The Texinfo/info/HTML/PDF documentation is the
authoritative source (try \fBinfo ne\fR or \fBinfo \-f ne\fR).
.SS OPTIONS
.TP
.I "--help"
Prints a help message.
.TP
.I "--"
Next token is a filename.
May appear more than once.
.TP
.I "+[N[,M]]"
Moves to the last or N-th line, first or M-th column of the next named file.
May appear more than once.
.TP
.TP
.I "--binary"
Load the next named file in binary mode.
May appear more than once.
.TP
.I "--read-only"
Load the next file in read-only mode.
May appear more than once.
.TP
.I "--utf8"
Use UTF-8 I/O.
.TP
.I "--no-utf8"
Do not use UTF-8 I/O.
.TP
.I "--ansi"
Use the built-in ANSI sequences.
.TP
.I "--no-ansi"
Never use the built-in ANSI sequences.
.TP
.I "--no-config"
Skip reading the menu and keyboard configuration files.
.TP
.I "--no-syntax"
Disable syntax-highlighting support.
.TP
.I "--prefs ext"
Set autoprefs for the provided extension before loading the first file.
.TP
.I "--keys key-configuration-file"
Use the specified keyboard configuration file.
.TP
.I "--menus menu-configuration-file"
Use the specified menu configuration file.
.TP
.I "--macro macro-file"
Execute the given macro after startup.
.SS USAGE
Start \fBne\fR, then use escape, escape-escape or F1 to access the menus.
.SS BUGS
Please send bug reports to Sebastiano Vigna <vigna@di.unimi.it> or Todd Lewis <utoddl@gmail.com>.
.SS AUTHORS
\fBne\fR was originally written by Sebastiano Vigna. Todd M. Lewis added
several new features. Daniele Filaretti helped with syntax highlighting.


NEWS
(This file, NEWS, lists new features and enhancements. See CHANGES for fixes.)

3.3.2 2022-09-13

  * Paste and PasteVert now set two bookmarks designated '<' and '>' which
    are set to the start and end of the most recently pasted text. Accessed
    by the "GotoBookmark <" and "GotoBookmark >" commands. Two associated
    menu items have been added to the "Search" menu.

  * Added toml.jsf syntax.

  * Support '_' in numeric constants, binary and hex constants, imaginary
    numbers in go.jsf syntax.

  * Now there is an uninstall target that uninstalls ne.

  * Additionally load key bindings from ~/.ne/.keys-${TERM}

3.3.1 2021-05-16

  * The makefile has been made more distro friendly by making it possible
    to override all compiler and linker options.

  * A workaround has been implemented to make --ansi (mostly) work when
    compiling with ncurses. A change to ncurses in 2012 made tgoto() not
    working unless tgetent()/setupterm() are called first.

3.3.0 2020-02-27

  * You can now Save (^A) named documents from within the
    SelectDoc (F4) requester.

  * The mark, if set, is visible.

  * Warning about opening a document with the same name as an existing
    document now considers the fully qualified canonical path and name
    of each rather than just their basenames.

  * Status bar Modified flag (*) may also be underlined, indicating the
    corresponding file's modification time has changed since the document
    was loaded or saved.

3.2.0 2019-09-30

  * Macros: They now work across multiple documents. You can cancel macro
    recording, and append new recorded commands to your current macro.

  * New documents inherit their predecessor's current macro, search, and
    replace strings.

  * Saving of read-only documents requires confirmation.

  * Word completion now does not consider the word the cursor is currently
    in, which helps in disambiguating.

  * The option -h is now an alias for --help. Thanks to Lonnie Abelbeck for
    implementing this feature.

  * Add asterisk.jsf syntax.

  * FIND in string inputs brings up a requester with your input history.
    Use Enter to replace, TAB to insert into your input line; Escape to exit.

  * Enhanced commands: MoveEOW, NextWord, and PrevWord can now move to either
    end of the indicated word via optional '<' and '>' parameter modifiers.

  * Requesters make better use of white space.

3.1.2 2018-10-06

  * RepeatLast now accepts "Find" or "Replace" after its optional number so
    you can explicitly repeat either operation even if the other was most
    recently performed.

  * We now use the xenl property to detect whether we can write on the last
    column of the last line, thus making the status bar complete.

3.1.1 2017-06-04

  * You can now CloseDoc (^Q) unmodified documents from within the
    SelectDoc (F4) requester.

3.1.0 2017-04-29

  * Large files that do not fit into memory are memory-mapped transparently
    if enough disk space is available.

  * Determine a "virtual extension" based on document contents and patterns
    from ~/.ne/.extensions. Allows overrides of specified extensions only.
    If no match, use .default#ap prefs.

  * Status flag "!" indicates the last line of document is not empty, thus
    the last line of the resulting saved file will not be terminated.

  * Requesters by default no longer remove non-matching entries as you
    enter characters. Use Insert or Delete keys within a requester
    to toggle between these behaviors.

  * Documents loaded through file requesters now have relative instead of
    absolute paths.

  * New NameConvert command (NC for short) changes current document's file
    name from relative to absolute path or absolute to relative path.

  * "SetBookmark ?" and "GotoBookmark ?" now prompt for the bookmark you
    want to set/goto. The prompt also indicates which bookmarks are set.

  * Command line option "--read-only" opens next document in read-only mode.

  * Options "--read-only" and "+N,M" can be applied to piped input:
      ls -l | ne --read-only +3,5 -

  * Warn if SaveAs would overwrite an existing file.

  * Suspend (Crtl-Z) works in prompts, requesters.

  * "KeyCode" now takes optional parameter.

  * Immediately/only after "Not Found", RepeatLast wraps around to the
    other end of the document.

3.0  2015-06-18

  * ne is now fully 64-bit, and needs to be compiled by a C99-compliant
    compiler. Files can be of any size, provided that enough core memory
    is available.

  * ne is able to read from named pipes. You can even pipe content into
    ne: it will be opened as the first document.

  * It is now possible to interrupt searches (it used to be possible just
    to interrupt global replace actions).

  * Word wrap has been reverted to pre-2.5: it preserves just whitespace,
    and it doesn't reparagraph at each line split or join. Paragraph keeps
    the usual smart features preserving comments, etc.

2.6  2015-04-17

  * Requesters (filenames, help, autocompletions, etc.) do progressive
    "fuzzy match" by typed characters, backspace.

  * Document requester (F4) opens with cursor on the current document.
    Docs with unsaved changes are bold, flagged with "*". F2/F3 reorder
    documents.

  * Tab in Syntax command displays requester of extant syntax recognizers.

  * You can now use AutoComplete in command line prompts.

  * New SaveAll command saves all modified documents.

  * Warns before saving over a file that was modified since the buffer was
    last loaded or saved.

  * `' joins (), {}, [], and <> as character pairs known to MatchBracket.

  * New example macros: aspell, DeleteSOL.

  * Add % to special leading characters for re-wrapping Paragraphs.

  * WordWrap no longer waits for the cursor to hit the right margin; it wraps
    on insertions and deletions now.

  * Updates to ruby, erb, python, conf, many other syntax recognizers.
    Added these recognizers from the Joe project: ant batch

src/ne.c
/* main(), global initialization and global buffer functions.

   Copyright (C) 1993-1998 Sebastiano Vigna
   Copyright (C) 1999-2022 Todd M. Lewis and Sebastiano Vigna

   This file is part of ne, the nice editor.

   This library is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3 of the License, or (at your
   option) any later version.

   This library is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
   or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
   for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, see <http://www.gnu.org/licenses/>.  */


#include "regex.h"

#include "ne.h"

#include "version.h"
#include "termchar.h"

#include <signal.h>
#include <limits.h>
#include <locale.h>

/* This is the array containing the "NO WARRANTY" message, which is displayed
   when ne is called without any specific file name or macro to execute. The
   message disappears as soon as any key is typed. */

char *NO_WARRANTY_msg[] = {   PROGRAM_NAME " " VERSION ".",
										"Copyright (C) 1993-1998 Sebastiano Vigna",
										"Copyright (C) 1999-2022 Todd M. Lewis and Sebastiano Vigna",
										"",
										"This program is free software; you can redistribute it and/or modify it under",
										"the terms of the GNU General Public License as published by the Free Software",
										"Foundation; either version 3 of the License, or (at your option) any later",
										"version.",
										"",
										"This program is distributed in the hope that it will be useful, but WITHOUT ANY",
										"WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A",
										"PARTICULAR PURPOSE.  See the GNU General Public License for more details.",
										"",
										"You should have received a copy of the GNU General Public License along with",
										"this program; if not, see <http://www.gnu.org/licenses/>.",
										"",
										"Press F1, Escape-Escape or Escape to see the menus. The shortcuts prefixed by ^",
										"are activated by the Control key; the shortcuts prefixed by [ are activated by",
										"Control+Meta or just Meta, depending on your terminal emulator. Alternatively,",
										"just press Escape followed by a letter.",
										"",
										"ne home page: http://ne.di.unimi.it/  GitHub repo: https://github.com/vigna/ne/",
										"Discuss ne at http://groups.google.com/group/niceeditor/",
										NULL
									};


char ARG_HELP[] = ABOUT_MSG "\n"
						"Usage: ne [options] [files]\n"
						"--help        print this message. [-h]\n"
						"--           *next token is a filename.\n"
						"+[N[,M]]     *move to last or N-th line, first or M-th column of next named file.\n"
						"--binary     *load the next file in binary mode.\n"
						"--read-only  *load the next file in read-only mode. [--readonly|--ro]\n"
						"--utf8        use UTF-8 I/O.\n"
						"--no-utf8     do not use UTF-8 I/O.\n"
						"--ansi        use built-in ANSI control sequences.\n"
						"--no-ansi     do not use built-in ANSI control sequences. [--noansi]\n"
						"--no-config   do not read configuration files. [--noconfig]\n"
						"--no-syntax   disable syntax-highlighting support.\n"
						"--prefs EXT   set autoprefs for the provided extension before loading the first file.\n"
						"--keys FILE   use this file for keyboard configuration.\n"
						"--menus FILE  use this file for menu configuration.\n"
						"--macro FILE  exec this macro after start.\n\n"
						"             *These options may appear multiple times.\n";


/* The regular expression used to parse the locale. */

#define LOCALE_REGEX "\\(UTF-?8\\)\\|\\(ISO-?8859-?\\)\\(1?[0-9]\\)"

/* These lists contain the existing buffers, clips and macros.
   cur_buffer denotes the currently displayed buffer. */

list buffers = { (node *)&buffers.tail, NULL, (node *)&buffers.head };
list clips = { (node *)&clips.tail, NULL, (node *)&clips.head };
char_stream *recording_macro;
bool executing_macro;
/* global prefs, only saved in ~/.ne/.default#ap if their
   current settings differ from these defaults. Make sure these
   defaults match the conditionals in prefs.c:save_prefs(). */
#ifndef ALTPAGING
bool req_order;
#else
bool req_order = true;
#endif
bool fast_gui;
bool status_bar = true;
bool interactive_mode;
bool verbose_macros = true;
/* end of global prefs */

buffer *cur_buffer;
int turbo;
bool do_syntax = true;

/* Whether we are currently displaying an about message. */
static bool displaying_info;

/* These functions live here because they access cur_buffer. new_buffer()
   creates a new buffer, adds it to the buffer list, and assign it to
   cur_buffer. delete_buffer() destroys cur_buffer, and makes the previous or
   next buffer the current buffer, if any of the two exists. */

buffer *new_buffer(void) {
	buffer *b = alloc_buffer(cur_buffer);

	if (b) {
		clear_buffer(b);
		if (cur_buffer) add(&b->b_node, &cur_buffer->b_node);
		else add_head(&buffers, &b->b_node);
		cur_buffer = b;
	}

	return b;
}

bool delete_buffer(void) {
	buffer *b = (buffer *)cur_buffer->b_node.next;

	rem(&cur_buffer->b_node);
	free_buffer(cur_buffer);

	if (! b->b_node.next) b = (buffer *)buffers.head;
	if (b == (buffer *)&buffers.tail) return false;

	cur_buffer = b;
	return true;
}

void about(void) {
	set_attr(0);
	clear_entire_screen();
	displaying_info = true;
	int i;
	for(i = 0; NO_WARRANTY_msg[i]; i++) {
		if (i == ne_lines - 1) break;
		move_cursor(i, 0);
		output_string(NO_WARRANTY_msg[i], false);
	}
	reset_window();

	char t[256] = ABOUT_MSG " Global directory";
	char * const gprefs_dir = exists_gprefs_dir();
	if (gprefs_dir) strncat(strncat(t, ": ", sizeof t - 1), gprefs_dir, sizeof t - 1);
	else strncat(strncat(strncat(t, " ", sizeof t - 1), get_global_dir(), sizeof t - 1), " not found!", sizeof t - 1);
	print_message(t);
}

/* The main() function. It is responsible for argument parsing, calling
   some terminal and signal initialization functions, and entering the
   event loop. */

int main(int argc, char **argv) {

	char *locale = setlocale(LC_ALL, "");
	for(int i = 0; i < 256; i++) localised_up_case[i] = toupper(i);

	if (locale) {
		struct re_pattern_buffer re_pb;
		struct re_registers re_reg;
		memset(&re_pb, 0, sizeof re_pb);
		memset(&re_reg, 0, sizeof re_reg);

		re_pb.translate = localised_up_case;
		re_compile_pattern(LOCALE_REGEX, strlen(LOCALE_REGEX), &re_pb);
		if (re_search(&re_pb, locale, strlen(locale), 0, strlen(locale), &re_reg) >= 0) {
			if (re_reg.start[1] >= 0) io_utf8 = true;
		}
		free(re_reg.start);
		free(re_reg.end);
	}

	bool no_config = false;
	char *macro_name = NULL, *key_bindings_name = NULL, *menu_conf_name = NULL, *startup_prefs_name = DEF_PREFS_NAME;

	char * const skiplist = calloc(argc, 1);
	if (!skiplist) exit(1);  /* We need this many flags. */

	for(int i = 1; i < argc; i++) {

		if (argv[i][0] == '-' && (!strcmp(&argv[i][1], "h") || !strcmp(&argv[i][1], "-help" "\0" VERSION_STRING))) {
			puts(ARG_HELP);
			exit(0);
		}

		/* Special arguments start with two dashes. If we find one, we
		   cancel its entry in argv[], so that it will be skipped when opening
		   the specified files. The only exception is +N for skipping to the
		   N-th line. */

		if (argv[i][0] == '-' && argv[i][1] == '-') {
			if (!argv[i][2]) i++; /* You can use "--" to force the next token to be a filename */
			else if (!strcmp(&argv[i][2], "noconfig") || !strcmp(&argv[i][2], "no-config")) {
				no_config = true;
				skiplist[i] = 1; /* argv[i] = NULL; */
			}
			else if (!strcmp(&argv[i][2], "noansi") || !strcmp(&argv[i][2], "no-ansi")) {
				ansi = false;
				skiplist[i] = 1; /* argv[i] = NULL; */
			}
			else if (!strcmp(&argv[i][2], "no-syntax")) {
				do_syntax = false;
				skiplist[i] = 1; /* argv[i] = NULL; */
			}
			else if (!strcmp(&argv[i][2], "prefs")) {
				if (i < argc-1) {
					startup_prefs_name = argv[i+1];
					skiplist[i] = skiplist[i+1] = 1; /* argv[i] = argv[i+1] = NULL; */
				}
			}
			else if (!strcmp(&argv[i][2], "ansi")) {
				ansi = true;
				skiplist[i] = 1; /* argv[i] = NULL; */
			}
			else if (!strcmp(&argv[i][2], "utf8")) {
				io_utf8 = true;
				skiplist[i] = 1; /* argv[i] = NULL; */
			}
			else if (!strcmp(&argv[i][2], "no-utf8")) {
				io_utf8 = false;
				skiplist[i] = 1; /* argv[i] = NULL; */
			}
			else if (!strcmp(&argv[i][2], "macro")) {
				if (i < argc-1) {
					macro_name = argv[i+1];
					skiplist[i] = skiplist[i+1] = 1; /* argv[i] = argv[i+1] = NULL; */
				}
			}
			else if (!strcmp(&argv[i][2], "keys")) {
				if (i < argc-1) {
					key_bindings_name = argv[i+1];
					skiplist[i] = skiplist[i+1] = 1; /* argv[i] = argv[i+1] = NULL; */
				}
			}
			else if (!strcmp(&argv[i][2], "menus")) {
				if (i < argc-1) {
					menu_conf_name = argv[i+1];
					skiplist[i] = skiplist[i+1] = 1; /* argv[i] = argv[i+1] = NULL; */
				}
			}
		}
	}

#ifdef NE_TEST
	{
		/* Dump the builtin menu and key bindings to compare to
		   doc/default.menus and doc/default.keys. */
		int dump_menu_config(FILE *f);
		int dump_key_config(FILE *f);
		FILE *f;

		if ((f = fopen("ne_test_dump_default_config", "w")) ) {
			dump_menu_config(f);
			dump_key_config(f);
			fclose(f);
		}
	}
#endif

	/* Unless --noconfig was specified, we try to configure the
	   menus and the keyboard. Note that these functions can exit() on error. */

	if (!no_config) {
		get_menu_configuration(menu_conf_name);
		get_key_bindings(key_bindings_name);
	}

#ifdef NE_TEST
	{
		/* Dump the builtin menu and key bindings to compare to
		   doc/default.menus and doc/default.keys. */
		int dump_menu_config(FILE *f);
		int dump_key_config(FILE *f);
		FILE *f;

		if ((f = fopen("ne_test_dump_loaded_config", "w")) ) {
			dump_menu_config(f);
			dump_key_config(f);
			fclose(f);
		}
	}
#endif

	/* If we cannot even create a buffer, better go... */

	if (!new_buffer()) exit(1);

	/* Now that key_bindings are loaded, try to fix up the
	   NOT_FOUND error_msg and the LONG_INPUT_HELP info_msg. */
	{
		char *keystroke_string, *new_msg_text;
		if ((keystroke_string = find_key_strokes(REPEATLAST_A, 1))) {
			if ((new_msg_text = malloc(39+strlen(keystroke_string)))) {
				strcat(strcat(strcpy(new_msg_text, "Not Found. (RepeatLast with "), keystroke_string), " to wrap.)");
				error_msg[NOT_FOUND] = new_msg_text;
			}
			free(keystroke_string);
		}
		if ((keystroke_string = find_key_strokes(FIND_A, 1))) {
			if ((new_msg_text = malloc(24+strlen(keystroke_string)))) {
				strcat(strcat(strcpy(new_msg_text, " (browse history with "), keystroke_string), ")");
				info_msg[LONG_INPUT_HELP] = new_msg_text;
			}
			free(keystroke_string);
		}
	}

	clear_buffer(cur_buffer);

	/* The INT_MAX clip always exists, and it is used by the Through command. */

	clip_desc * const cd = alloc_clip_desc(INT_MAX, 0);
	if (!cd) exit(1);

	add_head(&clips, &cd->cd_node);

	/* General terminfo and cursor motion initialization. From here onwards,
	   we cannot exit() lightly. */

	term_init();

	/* We will be always using the last line for the status bar. */

	set_terminal_window(ne_lines-1);

	/* We read in all the key capabilities. */

	read_key_capabilities();

	/* Some initializations of other modules... */

	re_set_syntax(
		RE_CONTEXT_INDEP_ANCHORS |
		RE_CONTEXT_INDEP_OPS     | RE_HAT_LISTS_NOT_NEWLINE |
		RE_NEWLINE_ALT           | RE_NO_BK_PARENS          |
		RE_NO_BK_VBAR            | RE_NO_EMPTY_RANGES
	);

	bool first_file = true;

	load_virtual_extensions();
	load_auto_prefs(cur_buffer, startup_prefs_name);

	buffer *stdin_buffer = NULL;
	if (!isatty(fileno(stdin))) {
		first_file = false;
		const int error = load_fd_in_buffer(cur_buffer, fileno(stdin));
		print_error(error);
		stdin_buffer = cur_buffer;

		if (!(freopen("/dev/tty", "r", stdin))) {
			fprintf(stderr, "Cannot reopen input tty\n");
			abort();
		}
	}

	/* The terminal is prepared for interactive I/O. */

	set_interactive_mode();

	clear_entire_screen();

	/* This function sets fatal_code() as signal interrupt handler
	   for all the dangerous signals (SIGILL, SIGSEGV etc.). */

	set_fatal_code();

	if (argc > 1) {

		/* The first file opened does not need a NEWDOC_A action. Note that
		   file loading can be interrupted (wildcarding can sometimes produce
		   unwanted results). */

		uint64_t first_line = 0, first_col = 0;
		bool binary = false, skip_plus = false, read_only = false;
		stop = false;

		for(int i = 1; i < argc && !stop; i++) {
			if (argv[i] && !skiplist[i]) {
				if (argv[i][0] == '+' && !skip_plus) {       /* looking for "+", or "+N" or "+N,M"  */
					uint64_t tmp_l = INT64_MAX, tmp_c = 0;
					char *d;
					errno = 0;
					if (argv[i][1]) {
						if (isdigit((unsigned char)argv[i][1])) {
							tmp_l = strtoll(argv[i]+1, &d, 10);
							if (!errno) {
								if (*d) {  /* separator between N and M */
									if (isdigit((unsigned char)d[1])) {
										tmp_c = strtoll(d+1, &d, 10);
										if (*d) errno = ERANGE;
									}
									else errno = ERANGE;
								}
							}
						}
						else errno = ERANGE;
					}
					if (!errno) {
						first_line = tmp_l;
						first_col  = tmp_c;
					}
					else {
						skip_plus = true;
						i--;
					}
				}
				else if (!strcmp(argv[i], "--binary")) {
					binary = true;
				}
				else if (!strcmp(argv[i], "--read-only") || !strcmp(argv[i], "--readonly") || !strcmp(argv[i], "--ro")) {
					read_only = true;
				}
				else {
					if (!strcmp(argv[i], "-") && stdin_buffer) {
						stdin_buffer->opt.binary = binary;
						if (read_only) stdin_buffer->opt.read_only = read_only;
						if (first_line) do_action(stdin_buffer, GOTOLINE_A, first_line, NULL);
						if (first_col)  do_action(stdin_buffer, GOTOCOLUMN_A, first_col, NULL);
						stdin_buffer = NULL;
					}
					else {
						if (!strcmp(argv[i], "--")) i++;
						if (!first_file) do_action(cur_buffer, NEWDOC_A, -1, NULL);
						else first_file = false;
						cur_buffer->opt.binary = binary;
						if (i < argc) do_action(cur_buffer, OPEN_A, 0, str_dup(argv[i]));
						if (first_line) do_action(cur_buffer, GOTOLINE_A, first_line, NULL);
						if (first_col)  do_action(cur_buffer, GOTOCOLUMN_A, first_col, NULL);
						if (read_only) cur_buffer->opt.read_only = read_only;
					}
					first_line =
					first_col  = 0;
					skip_plus  =
					binary    =
					read_only  = false;
				}
			}
		}

		free(skiplist);

		/* This call makes current the first specified file. It is called
		   only if more than one buffer exist. */

		if (get_nth_buffer(1)) do_action(cur_buffer, NEXTDOC_A, -1, NULL);

	}

	/* We delay updates. In this way the macro activity does not cause display activity. */

	#ifndef NE_TERMCAP
	if (ansi)
	#endif
	ttysize();
	reset_window();
	delay_update();

	if (macro_name) do_action(cur_buffer, MACRO_A, -1, str_dup(macro_name));
	else if (first_file) {
		/* If there is no file to load, and no macro to execute, we display
		   the "NO WARRANTY" message. */
		about();
	}

	while(true) {
		/* If we are displaying the "NO WARRANTY" info, we should not refresh the
		   window now */
		if (!displaying_info) {
			refresh_window(cur_buffer);
			if (!cur_buffer->visible_mark.shown) highlight_mark(cur_buffer, true);
			if (cur_buffer->opt.automatch) automatch_bracket(cur_buffer, true);
		}

		draw_status_bar();
		move_cursor(cur_buffer->cur_y, cur_buffer->cur_x);

		int c = get_key_code();

		if (window_changed_size) {
			print_error(do_action(cur_buffer, REFRESH_A, 0, NULL));
			window_changed_size = displaying_info = false;
			cur_buffer->automatch.shown = false;
			cur_buffer->visible_mark.shown = false;
		}

		if (c == INVALID_CHAR) continue; /* Window resizing. */
		const input_class ic = CHAR_CLASS(c);

		if (displaying_info) {
			refresh_window(cur_buffer);
			displaying_info = false;
		}

		if (cur_buffer->automatch.shown) automatch_bracket(cur_buffer, false);

		switch(ic) {
		case INVALID:
			print_error(INVALID_CHARACTER);
			break;

		case ALPHA:
			print_error(do_action(cur_buffer, INSERTCHAR_A, c, NULL));
			break;

		case TAB:
			print_error(do_action(cur_buffer, INSERTTAB_A, 1, NULL));
			break;

		case RETURN:
			print_error(do_action(cur_buffer, INSERTLINE_A, -1, NULL));
			break;

		case COMMAND:
			if (c < 0) c = -c - 1;
			if (key_binding[c]) print_error(execute_command_line(cur_buffer, key_binding[c]));
			break;

		default:
			break;
		}
	}
}
